### PR TITLE
DateBox: Fixing alphanumeric key behavior for setting time to "AM" and "PM" (T1216937)

### PR DESCRIPTION
**Pull request by jdvictoria**

### CODE REVIEW

**Comment by jdvictoria** [comment]
[Datebox - AM/PM can not be set by keyboard [T1216937]](https://trello.com/c/eOI4RzLO/8513-datebox-am-pm-can-not-be-set-by-keyboard-t1216937)

File: packages/devextreme/js/ui/date_box/ui.date_box.mask.js [hunk 1]
```diff
@@ -312,6 +312,11 @@ const DateBoxMask = DateBoxBase.inherit({
         const startString = this._searchValue + char.toLowerCase();
         const endLimit = limits.max - limits.min;
 
+        if(this._getActivePartProp('text') === 'AM' || this._getActivePartProp('text') === 'PM') {
```
Comment (AlexanderMoiseev) [inline packages/devextreme/js/ui/date_box/ui.date_box.mask.js]: 1) let's introduce a constant for this._getActivePartProp('text') and pass it to _handleAmPmSearch function.
2) i believe it os possible to move your code higher, before constants declarations. 
3) after refactoring i suggest getting rid of _handleAmPmSearch function, or at least renaming it

File: packages/devextreme/js/ui/date_box/ui.date_box.mask.js [hunk 2]
```diff
@@ -331,6 +336,18 @@ const DateBoxMask = DateBoxBase.inherit({
         }
     },
 
+    _handleAmPmSearch(char) {
+        const startString = this._searchValue + char.toLowerCase();
+
+        this._searchValue = startString;
```
Comment (AlexanderMoiseev) [inline packages/devextreme/js/ui/date_box/ui.date_box.mask.js]: please explain why did you add this code? it is not clear for me. if i remove it all tests pass

File: packages/devextreme/js/ui/date_box/ui.date_box.mask.js [hunk 3]
```diff
@@ -331,6 +336,18 @@ const DateBoxMask = DateBoxBase.inherit({
...
+
+        if(char.toLowerCase() === 'p' && this._getActivePartProp('text') === 'AM') {
+            this._toggleAmPm();
+        } else if(char.toLowerCase() === 'a' && this._getActivePartProp('text') === 'PM') {
+            this._toggleAmPm();
+        }
```
Comment (AlexanderMoiseev) [inline packages/devextreme/js/ui/date_box/ui.date_box.mask.js]:
```suggestion
if ((char === 'p' && activePartText === 'am') || (char === 'a' && activePartText === 'pm')) {
    this._toggleAmPm();
}
```
activePartText can be passed as a function parameter, plz look at my commets above.

File: packages/devextreme/js/ui/date_box/ui.date_box.mask.js [hunk 4]
```diff
@@ -312,6 +312,11 @@ const DateBoxMask = DateBoxBase.inherit({
         const startString = this._searchValue + char.toLowerCase();
         const endLimit = limits.max - limits.min;
 
+        if(this._getActivePartProp('text') === 'AM' || this._getActivePartProp('text') === 'PM') {
+            this._handleAmPmSearch(char);
```
Comment (AlexanderMoiseev) [inline packages/devextreme/js/ui/date_box/ui.date_box.mask.js]: lets pass char.toLowerCase()

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 5]
```diff
@@ -5760,6 +5760,48 @@ QUnit.module('datebox validation', {}, () => {
     });
 });
 
+[
```
Comment (AlexanderMoiseev) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: we have separate file for masked datebox tests. I suggest moving test to datebox.mask.tests. Also, do we really need to cretea a separate module for 1 test?

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 6]
```diff
@@ -5760,6 +5760,48 @@ QUnit.module('datebox validation', {}, () => {
...
+].forEach(({ type, input, displayFormat, expectedValue }) => {
+    QUnit.module(`DateBox AM/PM Handling - ${type} datebox type - ${displayFormat} a display format`, {
+        beforeEach: function() {
+            this.$element = $('#dateBox').dxDateBox({
+                value: new Date('10/10/2012 13:07'),
+                useMaskBehavior: true,
+                type: type,
+                mode: 'text',
+                displayFormat: displayFormat,
+                pickerType: 'calendar'
```
Comment (AlexanderMoiseev) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: please check if all these options are required for testing current scenario. eg pickerType

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 7]
```diff
@@ -5760,6 +5760,48 @@ QUnit.module('datebox validation', {}, () => {
...
+            this.instance = this.$element.dxDateBox('instance');
+            this.$input = this.$element.find('.dx-texteditor-input');
+            this.keyboard = keyboardMock(this.$input, true);
+        },
+        afterEach: function() {
+            fx.off = false;
+        }
+    }, () => {
+        QUnit.test(`when "${input}" is pressed it AM/PM should ${expectedValue ? `toggle to ${expectedValue}` : 'not toggle'} `, function(assert) {
+            this.instance.option('displayFormat', 'a');
```
Comment (AlexanderMoiseev) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: Why do you set the displayFormat at runtime? I mean, we can initialize it right away

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 8]
```diff
@@ -5760,6 +5760,48 @@ QUnit.module('datebox validation', {}, () => {
     });
 });
 
+[
+    { type: 'time', input: 'p', displayFormat: 'HH:mm a', expectedValue: 'PM' },
+    { type: 'time', input: 'a', displayFormat: 'HH:mm a', expectedValue: 'AM' },
+    { type: 'time', input: 'x', displayFormat: 'HH:mm a', expectedValue: null },
+    { type: 'time', input: 'y', displayFormat: 'HH:mm a', expectedValue: null },
+    { type: 'time', input: 'z', displayFormat: 'HH:mm a', expectedValue: null },
+    { type: 'datetime', input: 'p', displayFormat: 'dd/MM/yyyy HH:mm a', expectedValue: 'PM' },
```
Comment (AlexanderMoiseev) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: why do you use displayFormat parameter here? in the beginning of your test you set it to 'a' anyway.             this.instance.option('displayFormat', 'a');

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 9]
```diff
@@ -5760,6 +5760,48 @@ QUnit.module('datebox validation', {}, () => {
...
+            fx.off = false;
+        }
+    }, () => {
+        QUnit.test(`when "${input}" is pressed it AM/PM should ${expectedValue ? `toggle to ${expectedValue}` : 'not toggle'} `, function(assert) {
+            this.instance.option('displayFormat', 'a');
+
+            this.keyboard.type(input);
+
+            const actualValue = (input === 'p' || input === 'a') ? expectedValue : this.$input.val();
+            assert.strictEqual(this.$input.val(), actualValue);
```
Comment (AlexanderMoiseev) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: i think we should compare input value with expected value from for tests parameter, why do you use 'actualValue' param here?

**Review by AlexanderMoiseev — CHANGES_REQUESTED** [review]

File: packages/devextreme/js/ui/date_box/ui.date_box.mask.js [hunk 10]
```diff
@@ -306,6 +306,8 @@ const DateBoxMask = DateBoxBase.inherit({
     _searchString(char) {
         if(!isNaN(parseInt(this._getActivePartProp('text')))) {
             return;
+        } else {
```
Comment (AlexanderMoiseev) [inline packages/devextreme/js/ui/date_box/ui.date_box.mask.js]: 'else'  is redundant here, lest get rid of that

File: packages/devextreme/js/ui/date_box/ui.date_box.mask.js [hunk 11]
```diff
@@ -306,6 +306,8 @@ const DateBoxMask = DateBoxBase.inherit({
...
+            this._toggleAmPmOnChar(char.toLowerCase(), this._getActivePartProp('text'));
```
Comment (AlexanderMoiseev) [inline packages/devextreme/js/ui/date_box/ui.date_box.mask.js]: if we toggled am/pm, we may return from the function, i suppose, as you did before

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js [hunk 12]
```diff
@@ -1034,6 +1034,56 @@ module('Search', setupModule, () => {
...
+        });
+
+        this.keyboard.press('x');
+        assert.strictEqual(this.$input.val(), 'AM');
+
+        this.keyboard.press('y');
+        assert.strictEqual(this.$input.val(), 'AM');
+    });
+
     test('Hour', function(assert) {
```
Comment (AlexanderMoiseev) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js]: if you want to test different chars it is better to pass them as qunit tests arguments, the code will be simplified and more reusable. and tests will only contain 1 assert

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js [hunk 13]
```diff
@@ -1034,6 +1034,46 @@ module('Search', setupModule, () => {
...
+        });
+
+        this.keyboard.press('p');
+        assert.strictEqual(this.$input.val(), 'PM');
+    });
+
+    [
+        { letter: 'x', initialValue: 'AM', expectedValue: 'AM' },
+        { letter: 'y', initialValue: 'AM', expectedValue: 'AM' },
+        { letter: 'x', initialValue: 'PM', expectedValue: 'PM' },
```
Comment (AlexanderMoiseev) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js]: we could get rid of expected value here since initialValie always will be the same as expectedValue for this tests

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js [hunk 14]
```diff
@@ -1034,6 +1034,46 @@ module('Search', setupModule, () => {
...
+        assert.strictEqual(this.$input.val(), 'PM');
+    });
+
+    [
+        { letter: 'x', expectedValue: 'AM' },
+        { letter: 'y', expectedValue: 'AM' },
+        { letter: 'x', expectedValue: 'PM' },
+        { letter: 'y', expectedValue: 'PM' },
+    ].forEach(({ letter, expectedValue }) => {
+        QUnit.test('when other character is pressed it should not toggle PM to AM', function(assert) {
```
Comment (AlexanderMoiseev) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js]: i suggest this name - 'AM/PM part should not change when keys other than 'a' or 'p' are pressed.'

Comment (jdvictoria) [reply packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js]: That's better

**Review by AlexanderMoiseev — DISMISSED** [review]

File: packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js [hunk 15]
```diff
@@ -1034,6 +1034,46 @@ module('Search', setupModule, () => {
         assert.strictEqual(this.$input.val(), 'PM', 'revert incorrect changes');
     });
 
+    test('when A is pressed it should toggle PM to AM', function(assert) {
```
Comment (AlexanderMoiseev) [inline packages/devextreme/testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js:1037]: please add ticket number to the test's name

**Review by AlexanderMoiseev — DISMISSED** [review]
